src/main/java/org/springframework/retry/support/MetricsRetryListener.java

```diff
@@ -0,0 +1,132 @@
+ * exception class name</li>
+ * </ul>
+ * <p>
+ * The {@link #setCustomTags(Map)} and {@link #setCustomTagsProvider(Function)} can be
+ * used to further customize tags on the timers.
+ *
+ * @author Artem Bilan
+ * @since 2.0.8
+ */
+public class MetricsRetryListener implements RetryListener {
+
+	public static final String TIMER_NAME = "spring.retry";
+
+	private final MeterRegistry meterRegistry;
+
+	private final Map<RetryCallback<?, ?>, Timer.Sample> retryContextToSample = new HashMap<>();
+
+	private final Map<String, String> customTags = new HashMap<>();
+
+	private final Meter.MeterProvider<Timer> retryMeterProvider;
+
+	private Function<RetryContext, Iterable<Tag>> customTagsProvider = retryContext -> Tags.empty();
+
+	/**
+	 * Construct an instance based on the provided {@link MeterRegistry}.
+	 * @param meterRegistry the {@link MeterRegistry} to use for timers.
+	 */
+	public MetricsRetryListener(MeterRegistry meterRegistry) {
+		Assert.notNull(meterRegistry, "'meterRegistry' must not be null");
+		this.meterRegistry = meterRegistry;
+		this.retryMeterProvider = Timer.builder(TIMER_NAME)
+			.description("Metrics for Spring RetryTemplate")
+			.withRegistry(this.meterRegistry);
+	}
+
+	/**
+	 * Supply a map of tags which are going to be used for all the timers managed by this
+	 * listener.
+	 * @param customTags the map of additional tags for all the timers.
+	 */
+	public void setCustomTags(Map<String, String> customTags) {
+		Assert.notNull(customTags, "'customTags' must not be null");
+		this.customTags.putAll(customTags);
+	}
+
+	/**
+	 * Supply a {@link Function} to build additional tags for all the timers based on the
+	 * {@link RetryContext}.
+	 * @param customTagsProvider the {@link Function} for additional tags with a
+	 * {@link RetryContext} scope.
+	 */
+	public void setCustomTagsProvider(Function<RetryContext, Iterable<Tag>> customTagsProvider) {
+		Assert.notNull(customTagsProvider, "'customTagsProvider' must not be null");
+		this.customTagsProvider = customTagsProvider;
+	}
+
+	@Override
+	public <T, E extends Throwable> boolean open(RetryContext context, RetryCallback<T, E> callback) {
+		this.retryContextToSample.put(callback, Timer.start(this.meterRegistry));
+		return true;
+	}
+
+	@Override
+	public <T, E extends Throwable> void close(RetryContext context, RetryCallback<T, E> callback,
+			Throwable throwable) {
+
+		Timer.Sample sample = this.retryContextToSample.remove(callback);
+
+		Assert.state(sample != null,
+				() -> String.format("No 'Timer.Sample' registered for '%s'. Was the 'open()' called?", context));
+
+		Tags retryTags = Tags.of("name", callback.getLabel())
+			.and("retry.count", "" + context.getRetryCount())
+			.and(this.customTags.entrySet().stream().map((entry) -> Tag.of(entry.getKey(), entry.getValue())).toList())
+			.and(this.customTagsProvider.apply(context));
+
+		if (throwable != null) {
+			retryTags = retryTags.and("exception", throwable.getClass().getSimpleName());
+		}
```

Comment:
This condition can make things tricky: it will mean that you can end up with time-series that has the same name but different set of tags (exception vs. no exception):
```
spring.retry{name=test, retry.count=0} ...
spring.retry{name=test, retry.count=0, exception=BOOM!} ...
```
This is usually discouraged but more than that, for the Prometheus Java client this is invalid and the metrics will be dropped. You can avoid this by adding a default value to the `exception` tag, for example `none` (MVC, WebFlux, etc. are doing this):
```
spring.retry{name=test, retry.count=0, exception=none} ...
spring.retry{name=test, retry.count=0, exception=BOOM!} ...
```
You can do something like this:
```java
private String exceptionTagValue(Throwable throwable) {
	if (throwable != null) {
		return throwable.getClass().getSimpleName();
	}
	else {
		return "none";
	}
}
```
and in `retryTags` simply:
```java
Tags retryTags = Tags.of("name", callback.getLabel())
	.and("retry.count", "" + context.getRetryCount())
	.and("exception", exceptionTagValue(throwable))
```

Reply:
Now I know why I see `exception=none` in `spring-kafka` and `spring-integration`😄